### DISCUSSION TITLE
How to change the Buffer size on the TCP socket on INET?

### DISCUSSION THREAD

**Discussion by GIANNIS-AGGELIS** [omnetpp/omnetpp#1061] [Q&A] [answered]
TCP is a complex protocol that provides lots of functionality for reliable transmission on the network. From the literature one of many things that can affect the performance of TCP is Bandwidth Delay Product (BDP). I have fallen into many papers that use the value BDP or multiple thereof for the buffer of TCP.

The first thought that I had was: What Buffer?
First, I believed that was the value of advertisedWindow from TCP but it was not. I ran different experiments with the same settings as the authors from those papers and I couldn't verify their results.

After some more digging, I found that they talk about the buffer size on the TCP socket for the sender and the receiver [1] (section 2.1). I also found in Linux Distros you can change values of Buffer SO_SNDBUF, SO_RCVBUF respectively tcp_wmem,tcp_rmem. 

Who you are so sure that are those buffers? 
Ans: From this course [2] (page 15)

The thing is that I don't know how to give a value to the buffer on INET.
Is there any solution to this problem or any suggestion??

I will add my network and the .ini file below. I try to recreate the simple Figure 1 for this [3] paper (page 3). I use the same parameters as the author's 400Mbps bandwidth one-way delay 180ms, with two Standard Hosts, and for the app, I use TcpSessionApp and TcpSinkApp.

As you can see in the scenario I do not lose any segment over 12329 segments which is the BDP of the link. I have included two plots from the vectors cwnd, unacked date. Also in the zip the source code for the simulation.

[cwnd Plot.pdf](https://github.com/omnetpp/omnetpp/files/13031098/cwnd.Plot.pdf)
[unacked data Plot.pdf](https://github.com/omnetpp/omnetpp/files/13031099/unacked.data.Plot.pdf)

[HyStart.zip](https://github.com/omnetpp/omnetpp/files/13031109/HyStart.zip)

[1] http://netlab.caltech.edu/FAST/references/net100.sc02.final.pdf
[2] http://ce.sc.edu/cyberinfra/workshops/Material/NTP/Lab%208.pdf 
[3] https://www.researchgate.net/profile/Sangtae-Ha/publication/228960212_Hybrid_slow_start_for_high-bandwidth_and_long-distance_networks/links/574b20e708ae5bf2e63f344d/Hybrid-slow-start-for-high-bandwidth-and-long-distance-networks.pdf

**Comment by avarga** [comment, marked as answer]
In my understanding, `tcp_wmem` in Linux describes the size of the socket's write buffer. If the sender process overflows the buffer, then the send call is blocked or fails with EAGAIN. Setting it higher should not affect TCP throughput, as long as TCP always has enough data to send. In INET we don't have this parameter, i.e. the sender app is never blocked, and socket writes never fail. As long as your app takes care to always write enough data into the socket to keep TCP fully busy, throughput will not be harmed.

The `tcp_rmem` (received buffer size) in Linux is the amount of received data the kernel is willing to buffer until the sink app reads it out of the socket. If the sink app does not read enough, and the free space in the receive buffer gets smaller than 64K (or 64K * 2^scaling if window scaling is on), then the receiver TCP starts advertising reduced windows, and flow control kicks it, reducing the throughput. In INET, currently we don't have a receive buffer size.  We haven't needed one, because until recently, TCP sockets operated in implicit "autoread" mode, i.e. received data were immediately sent up to the application. The inet `master` branch now we have support for explicit read requests, which can be turned on by setting `autoRead=false` in apps that support it. We still haven't added a `receiveBuffer` parameter, but we do have an `advertisedWindow` parameter on TCP. When you use it, make sure you turn on window scaling (on both endpoints, because it only takes effect then) and specify a large advertisedWindow (at least bandwidth*delay).

**Reply by GIANNIS-AGGELIS** [reply]
@avarga Thank you for this clarification you are right. I do not need buffers in sockets to test and implement the HyStart algorithm. I apologize for the late response.
